### ISSUE TITLE
Arrange How We'll Work Here and Collaborate With Upstream

### ISSUE THREAD

**Issue by RandomDSdevel** [portage-brew/brew#2]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I'm initially somewhat inclined to want to work off of upstream's latest stable released tag (by tracking a '`stable`' branch like what I have set up locally) and rebase as needed, but we could also work of of '`master`.'  Regardless, we need to find some way to make it easy to stay abreast of upstream commits, or at least most of them and/or the ones we'd be interested in.  I also want us to consider how we'd upstream commits if we decided to do so for any before fully upstreaming our work here.

**Comment by rw** [comment]
Short term, overriding `pour_bottle?` to always be `false` should be enough IMO.

Long term, their changes to building packages (no more options?!) for v2 might force us to diverge.

**Comment by RandomDSdevel** [comment]
@rw:  

> ⋮
>
> Long term, their changes to building packages (no more options?!) for v2 might force us to diverge.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;There would, at least initially, likely remain a non-trivial number of commits we could keep in common, however, and I'd want to make maintaining that `brew` code similarity tenable.  We'd handle differences in formulae between upstream's and our copies in an organization fork of [Homebrw/homebrew-core](https://github.com/Homebrew/homebrew-core), anyway, and, per portage-brew/portage-brew-staging-and-evolution#8, that doesn't exist yet.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;With respect to branch organization here in portage-brew/brew, here, tentatively, is what I have set up for now:  

- '`upstream-master`' tracks upstream's '`master`.'  
   - Our '`master`' branch tracks '`upstream-master`.'  
- '`upstream-stable`' tracks upstream's latest stable tagged release.  
   - Our '`stable`' branch tracks '`upstream-stable`.'  (Odds could be that this should track our '`master`' branch…?)  

I haven't quite hammered out whether our PRs should be against our '`master`' or our '`stable`' branch and would appreciate guidance on that.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An update:  I've reorganized which branches have which others as upstreams.  Here's the new tentative layout:  

- '`upstream-master`' tracks upstream's '`master`.'  
   - Our '`master`' branch tracks '`upstream-master`.'  
      - Our '`stable`' branch tracks our '`master`' branch.  
      - PRs are tentatively against our '`master`' branch.  
- '`upstream-stable`' tracks upstream's latest stable tagged release.  

Having our '`stable`' branch track our '`master`' branch instead of our '`upstream-stable`' branch comes with a trade-off:  

- It makes it easier to pull our downstream commits in from '`master`,' but more difficult to otherwise keep based off of a stable upstream release.  

   (Instead of only manually pulling the latest tag into '`upstream-stable`' and then rebasing '`stable`' on top of that, one instead has to manually pull the latest tag into '`upstream-stable`' _and_ separately rebase '`stable`' on top of that same upstream tag.)  

I'm uncertain how this interacts, if at all well, with pulling commits from our '`master`' branch into our '`stable`' branch.  Maybe somebody can suggest a cleaner solution here…

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Also, in case it wasn't clear, I want to have tooling and/or automation set up to ease maintaining whatever we settle on here.

**Comment by SMillerDev** [comment]
> Long term, their changes to building packages (no more options?!) for v2 might force us to diverge.

Options aren't going away in brew, just in the core repo.

**Comment by sjackman** [comment]
I'm the lead of maintainer of [Linuxbrew/brew](https://github.com/Linuxbrew/brew), which has been a fork of Homebrew/brew for about five years. Each stable release of Homebrew, I merge the stable tag (say `1.9.3`) into branch a branch named `merge-1.9.3` on Linuxbrew/brew. I open a pull request from that branch. When it passes CI, I merge that pull request (using `git checkout master && git merge --ff-only merge-1.9.3 && git push origin master` and delete the branch `merge-1.9.3`). Developers and our CI will now be using that `master` branch. I leave it for about a day, to wait for bug reports. After about a day, I push the stable tag (`1.9.3`) and release it.

**Comment by RandomDSdevel** [comment]
@sjackman:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;So you don't have any analog of our '`upstream-master`' branch?  You only keep track of stable upstream releases?

**Comment by sjackman** [comment]
Correct. We merge only stable upstream releases into our `master` branch.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;OK, that makes things simpler.  I'll implement that soon.

**Comment by sjackman** [comment]
`upstream-master`: I have a remote named `homebrew` and call this branch `homebrew/master`, which is updated with `git fetch homebrew`. No harm in giving it a pseudonym though.

**Comment by RandomDSdevel** [comment]
@sjackman:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Thought I'd seen that as a convention somewhere, codifying/mirroring the terminology.  Eh, it can always get renamed later to be more clear if need be.

**Comment by sjackman** [comment]
> No harm in giving it a pseudonym though.

So long as you configure `brew fetch` to automatically update `upstream-master` from its remote. Otherwise it's one more branch name to get stale and confuse you.

**Comment by RandomDSdevel** [comment]
@sjackman:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;You mean '`brew update`.'  '`brew fetch`' is for formula artifacts, as I recall.

**Comment by sjackman** [comment]
Typo. I meant `git fetch`. I swap `git` and `brew` multiple times per day. I may never get them straight.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;No worries; I do it sometimes, too!

**Comment by S-C-Johnston** [comment]
I thought it'd be more appropriate to comment here than start a new issue on this: I think it'd be wise to document some of the architecture of homebrew, either as a part of this project or as a part of another project in this organization. The chief impetus is that homebrew org doesn't seem to make public much of their design except in raw ruby. It seems an odd half-measure between cathedral and bazaar styles, there. Anyway, I say this because I think it'd help lower the barrier to contributing in a meaningful way.

At the very least, some simple `graphviz` diagrams (a good medium precisely because it is compiled from markup, and therefore put in git and collaborated on) that illustrate the interaction between the user and homebrew and source code.

I have some informed guesses I could make about its structure and interaction. For example, just sampling a few commits, it looks like the maintenance of most formula revolves around incrementing the version number and updating the hashes. That doesn't do a whole lot to tell me in which files I should read the relevant source, though.

---

I am a student with a full-time job commitment on top of that, so I'm not sure how much I could help, frankly. And perhaps it's wildly unhelpful to be spitballing as an idea-guy. I just know that it seems super intimidating from my perspective to even begin to understand where things should be worked on.

**Comment by sjackman** [comment]
I'm a maintainer of Homebrew. I'm happy to answer specific questions that you may have about the architecture of Homebrew. In the mean time, there's a fair amount of information in the docs: https://docs.brew.sh particularly https://docs.brew.sh/Formula-Cookbook
